Add a remote-close integration test.

mollyling opened the pull request:
Yeah, just:
-  setup to listen to a close event locally, 
-  tell the echo server to close its connections
- then make sure we got the close event.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy-networking/226)

<!-- Reviewable:end -->

bemasc committed:
Intermediate test state.  Crashes Chrome!!!

Commit:
Better request-size options in benchmark.

mollyling committed:
Add a remote-close integration test.

bemasc committed:
Implement port mirroring.

This allows CHURN connections to peers who are behind symmetric NAT.

https://github.com/uProxy/uproxy/issues/608

bemasc committed:
Implement load tests.

The tests all pass in normal mode, and all fail for CHURN.

bemasc committed:
Rename ProxyIntegrationTestClass to AbstractProxyIntegrationTest

iislucas committed:
Merge pull request #225 from uProxy/dev-requireify

Requireify of uproxy-networking

jpevarnek committed:
Update to new version of uproxy-lib

This updates the version of uproxy-lib, fixes our calls to the logger to
use the new format, and adds a logging listener to each freedom module
so we will get updates when levels change.

trevj committed:
fix paths for echo server sample on firefox

trevj committed:
fix paths for simple socks on firefox

trevj committed:
bump patch version for firefox samples fix

bemasc committed:
Improve clarity in response to review comments.

trevj committed:
Merge pull request #234 from uProxy/trevj-firefox-samples

make the firefox samples run again

trevj committed:
fix path in copypaste sample freedom manifest

trevj committed:
Merge pull request #235 from uProxy/trevj-copypaste-fix

fix path in copypaste sample freedom manifest

Commit:
update build instructions for setup.sh

iislucas committed:
fixes for dist files and integration test to not be interactive

iislucas committed:
Merge pull request #236 from uProxy/iislucas-dist-and-intergation-fixes

Make dist files be more sensible; make integration test to not be interactive

jpevarnek committed:
Merge pull request #232 from uProxy/jpevarnek-update-uproxy-lib

Update to new version of uproxy-lib

iislucas committed:
fix to avoid grunt-jasmine-chromeapp race condition on its exit with copy

iislucas committed:
Merge pull request #237 from uProxy/iislucas-gruntfile

Gruntfile tweak

trevj committed:
always set freedom debug level to debug

trevj committed:
Merge pull request #238 from uProxy/trevj-freedom-debug

always set freedom debug level to debug

mollyling commented:
The continuous build failure is on: "Warning: Unable to read "/home/travis/build/uProxy/uproxy-networking/node_modules/freedom-pgp-e2e/build/*_/_" file (Error code: ENOENT). Use --force to continue.".  Which frankly, befuddles me.  Should I rebase off of a newer branch to fix?

bemasc committed:
Manage data channels as a pool, to avoid issues with channel closing.

The CHURN integration tests now pass, so they are now enabled
by default.

BUG=https://github.com/uProxy/uproxy/issues/1309

Commit:
simplifying build instructions

Commit:
readme build path

Commit:
Merge pull request #239 from uProxy/soycode-buildinstructions

update build instructions for setup.sh

mollyling committed:
Merge pull request #207 from uProxy/mling-update-benchmark

Better request-size options in benchmark.

bemasc committed:
Move domain names to debug level logging.

iislucas committed:
fix bad style for freedom module creation

iislucas commented:
Yay! More integration tests!

That's probably just an old `freedom-pgp-e2e` library in node_modules; nuke the node modules and re-install. 

Also: you probably want to redo this pull request against dev (dev requirefiy is now merged into dev)

bemasc committed:
Move the whole log message to debug.

bemasc committed:
Merge pull request #244 from uProxy/bemasc-hide-domains

Move domain names to debug level logging.

bemasc committed:
Merge pull request #241 from uProxy/bemasc-pool

Never close an RTCDataChannel

trevj committed:
add SocketCloseKind to TCP server onceClosed and lots of logging and formatting cleanup

trevj committed:
add space

iislucas committed:
Merge branch 'dev' into iislucas-cleanup

iislucas committed:
tiny fix to make work - missing loggingController prefix

trevj committed:
use default value for constructor arg

trevj committed:
use socketId for array index

trevj committed:
prepend a unique instance ID to server socket logging statements

trevj committed:
use socketId for array index here, too

trevj committed:
log server socket close kind, remove unneeded catch block

trevj committed:
remove isListening and isShutdown methods

trevj committed:
add ID to server toString and tidy a little

trevj committed:
tcp server now returns SocketCloseKind

bemasc committed:
Remove obsolete SocksToRtc.sessions_ attribute

bemasc committed:
Rename "tag" to "channelLabel" and minor cleanup.

bemasc committed:
Merge pull request #248 from uProxy/bemasc-pool

Remove obsolete SocksToRtc.sessions_ attribute

trevj committed:
Merge pull request #247 from uProxy/trevj-tcp-server-cleanup

add SocketCloseKind to TCP server onceClosed, with cleanup

trevj committed:
move to heartbeats-based uproxy-lib

bemasc committed:
Merge pull request #249 from uProxy/trevj-heartbeats

move to heartbeats-based uproxy-lib

trevj committed:
fix paths to loggingprovider in firefox samples

bemasc committed:
Make PoolChannels respect closure of the underlying DataChannel.

This is needed to ensure that cleanup actions run when the
PeerConnection terminates.

bemasc committed:
Respond to review comments

bemasc committed:
Remove superfluous noopPromise

bemasc committed:
Merge pull request #251 from uProxy/bemasc-pool

Make PoolChannels respect closure of the underlying DataChannel.

bemasc committed:
Add comment describing slow tests

bemasc committed:
Merge pull request #252 from uProxy/bemasc-load-test

Implement load tests

trevj committed:
Merge branch 'dev' into trevj-firefox-loggingprovider-path-fix

trevj committed:
Merge pull request #250 from uProxy/trevj-firefox-loggingprovider-path-fix

fix paths to loggingprovider in firefox samples

trevj commented on src/integration-tests/socks-echo/base-spec.core-env.ts:
```diff
@@ -54,6 +54,28 @@ export function socksEchoTestDescription(useChurn:boolean) {
+  it('detects a remote close', (done) => {
+    var input = arraybuffers.stringToArrayBuffer('arbitrary test string');
+    var testModule = getTestModule();
+    var connid : string;
```
super nit:

``` typescript
var connId :string;
```

trevj commented on src/integration-tests/socks-echo/base-spec.core-env.ts:
```diff
@@ -54,6 +54,28 @@ export function socksEchoTestDescription(useChurn:boolean) {
+    var input = arraybuffers.stringToArrayBuffer('arbitrary test string');
+    var testModule = getTestModule();
+    var connid : string;
+    var testModuleCopy:any;
```
This is weird, I don't see where it's used.

trevj commented:
Again, it would be good to rebase this off of the `dev` branch. Thanks again for adding this!

bemasc committed:
Prefer public host candidates over server-reflexive candidates.

This allows CHURN to work when one endpoint is a symmetric NAT,
and the other endpoint has both port-restricted-NAT and non-NAT
interfaces.

jpevarnek committed:
Add a DefinitelyTyped rule for circular-json

We have the really annoying dependency rule that we need to pull in
typings for any child repo since the files get copied to our third_party
directory instead of living in their build directories that have the
typings specified.

jpevarnek committed:
Update version to 10.0.2

jpevarnek committed:
Merge pull request #253 from uProxy/typings-circular-json

Add a DefinitelyTyped rule for circular-json

trevj committed:
always destroy socket provider on disconnect, and fix type error

trevj committed:
fix indentation

trevj committed:
simplify and shorten comment for destroyFreedomSocket_

bemasc committed:
Document the port selection heuristic.

bemasc committed:
Merge pull request #233 from uProxy/bemasc-mirror-merge

Implement port mirroring.

trevj committed:
track number of inflight calls to the socket and only destroy its provider when all have finished

trevj committed:
also have client sockets track their number of inflight calls for socket destruction

trevj committed:
simplify wrap async handling and throw before exceptions synchronously

trevj committed:
add comment for inflight termination condition

trevj committed:
sockets should be bound to sockets

trevj committed:
extract socket provider call counting to a separate class

trevj committed:
workaround for core.udpsocket.getInfo address reporting issue

trevj committed:
onDisconnect handlers should wait until the counter is destroyed to fulfill onceClosed

trevj committed:
explain how discard works

trevj committed:
Merge pull request #256 from uProxy/trevj-churn-firefox-fix

workaround for core.udpsocket.getInfo address reporting issue

trevj committed:
move wrap function to the counter class

trevj committed:
Merge remote-tracking branch 'origin/dev' into trevj-tcp-close

trevj committed:
bump patch version for tcp socket closing fix

trevj committed:
Merge branch 'dev' into trevj-tcp-close

trevj committed:
Merge pull request #254 from uProxy/trevj-tcp-close

always destroy socket provider on disconnect, and fix type error

bemasc committed:
Refactor churn-pipe for better efficiency

bemasc committed:
Churn-pipe cleanup in response to review comments.

bemasc committed:
Don't enable obfuscation by default for copypaste-socks just yet.

bemasc committed:
Merge pull request #258 from uProxy/bemasc-churn-module

Refactor churn-pipe for better efficiency

trevj committed:
Merge pull request #245 from uProxy/iislucas-cleanup

fix bad style for freedom module creation

mollyling committed:
@trevj changes.

mollyling commented:
<sup>**[src/integration-tests/socks-echo/base-spec.core-env.ts, line 60 [r1]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/226#-JmKV8fVS9IhPSFR8mYC-r1-60)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/21bfe3bb2727e8c6b86b7233382b175a7d02559b/src/integration-tests/socks-echo/base-spec.core-env.ts#L60)):</sup>
Done.

---

<sup>**[src/integration-tests/socks-echo/base-spec.core-env.ts, line 61 [r1]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/226#-JmKV8fVS9IhPSFR8mYC-r1-61)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/21bfe3bb2727e8c6b86b7233382b175a7d02559b/src/integration-tests/socks-echo/base-spec.core-env.ts#L61)):</sup>
Done.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-networking/226)

<!-- Sent from Reviewable.io -->

mollyling committed:
Cleanup and build updates.

mollyling commented:
Rebased to dev.

mollyling commented:
This has gone somewhere awful.  Killing it.
